docs/configuration.rst

```diff
@@ -89,6 +89,33 @@ ACCOUNT_EMAIL_MAX_LENGTH(=254)
   3-byte and 4-byte Unicode character sets
   <https://dev.mysql.com/doc/refman/5.5/en/charset-unicode-conversion.html>`_.
 
+ACCOUNT_TIMEOUTS(={})
+
+  Allows you to set a timeout (in seconds) on how often a user can
+  do a certain action.
+
+  e.g. by default the password reset form sends an email every single time
+  someone fills it out. Setting the timeout will prevent someone requesting
+  a password reset without waiting.
+
+  In this example, restrict the ResetPasswordForm to only allow sending an
+  email every 5 minutes::
+
+    ACCOUNT_TIMEOUTS = {
+        'ResetPasswordForm': 5*60,
+        'ResendEmailVerification': 5*60,
+        'ChangePasswordForm': 10,
```

Comment:
I think that the use case for these actions is solely limited to the sending of emails, right? The email is used as cache key after all. Though, the naming of the setting suggests that this is a more general concept ("certain action a user can do"). So, I am wondering if this setting should not be named ACCOUNT_EMAIL_TIMEOUTS.

At the same time, I am wondering how one would configure these values. If the goal is to prevent people from triggering too many mails, why do we need multiple individually configurable values instead of having just one general email cooldown value.

We now have a EMAIL_CONFIRMATION_COOLDOWN setting -- wouldn't renaming this into a general EMAIL_COOLDOWN setting and using it all over where emails are sent suffice?

Reply:
@pennersr 

Yes the sending of emails since all of those actions (reset password, resend email verification, change password) involve sending an email. 

That is not a bad idea to rename the ACCOUNT_TIMEOUTS setting to EMAIL_COOLDOWN where the one assigned value to that setting controls all 3 actions instead of keeping them separated. As long as this setting is separate from the EMAIL_CONFIRMATION_COOLDOWN setting that seems reasonable.

From a UX standpoint though, does it make sense to only provide 1 value to the EMAIL_COOLDOWN setting to control all 3 email actions? Would that one value for one email action make sense everywhere else for the other email actions? If not, then keep in mind an allauth programmer would need to change the one value for all 3 email action places which will then get affected all the same. The end user (site visitor) would need to wait that single allotted time for all 3 actions. 

I imagine an end user scenario like this - "I tried resending verification too many times and now I can't change my password I need to wait the same amount of time to do anything else". Versus "I tried resending verification too many times and now at least I can change my password but need to wait less time to do it". 

Not sure if the PR here works like that scenario above though.

Should one of the following things happen?
1) Site visitor can perform action A x number of times per minute. Then action A only gets timed out. Actions B and C do not get timed out until the user performs them. 
2) Site visitor can perform action A x number of times per minute. Then actions A, B, and C all get timed out for the same amount of time.
3) Something else? 

I am fine with either combining the actions or separating the actions into that EMAIL_COOLDOWN setting as 1 value or multiple values though. I leave that to you all to decide. 

@iarp 

Curious to hear your opinions on how this should be handled.

Reply:
At first it was written to be an email cooldown type system but as I wrote it up I realized it could be applied to various other number of things not just email, for instance limiting how fast someone can do a certain action, such as changing a password.

Reply:
@iarp Right, so are you leaning towards penners idea or keeping each action configured separately? From a UX standpoint, does it make sense to time out everything for the same time period as long as one action gets abused? I imagine users might not like that if that is the case because they have to wait to do the other actions but not 100% sure if that is what penners means. How does your code currently work? Does it timeout an action as long as that one action gets abused? No other action gets timed out if one is abused? Or does your code do something else?

Reply:
I'm starting to think this is growing beyond allauth's scope.

The adapter methods are written in a way that means they could be used for any sort of timeout purpose. Which is why I included the ChangePasswordForm as an example, I personally cannot find any reason within me that I would ever time that out but its an example of a non-email based thing happening with timeout restrictions.

It's setup to allow an admin to set custom timeouts because I think certain actions should be allowed faster than others. Resetting a users password via email requires time, the user needs to allow enough time for smtp connection, propagation, checking spam folders.. etc. But then if I use the changing password example, why restrict that to 5 minutes? 30 seconds should be long enough. The real question for someone hitting the change password timeout is why are they changing a password so fast like that?

Reply:
I agree with this. Sorry for making things more confusing.